src/java.base/share/classes/java/io/BufferedInputStream.java

```diff
@@ -643,9 +643,14 @@ private long implTransferTo(OutputStream out) throws IOException {
         if (getClass() == BufferedInputStream.class && markpos == -1) {
             int avail = count - pos;
             if (avail > 0) {
-                // Prevent poisoning and leaking of buf
-                byte[] buffer = Arrays.copyOfRange(getBufIfOpen(), pos, count);
-                out.write(buffer);
+                // trust all OutputStreams from java.io
+                if (out.getClass().getPackageName() == BufferedInputStream.class.getPackageName()) {
```

Comment:
I don't think Class::getPackageName documents that the returned String is intern so I wonder if the == check will lead to questions and suggestions of a bug. Classes with names starting with "java." can only be defined to the boot or platform class loader (details in the ClassLoader API docs) so you could just check if the package name equals "java.io".

Reply:
Do we only want to trust java.io or anything starting with java.*?

Reply:
I don't think checking if the package is java.io is secure:
```
ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
UntrustedOutputStream uos = new UntrustedOutputStream();
bais.transferTo(new java.io.DataOutputStream(uos)); 
```
You have to know that it is in the java.io package and it doesn't wrap another stream.

Reply:
@mkarg I've monitored the source code and it turns out that outside of `java.io` we have `Process`-related `NullOutputStream`, `SocketOutputStream`, `ZipOutputStream`, `InflaterOutputStream`, `DeflaterOutputStream`, `GZIPOutputStream`, `CheckedOutputStream`, `JarOutputStream`, `EncOutputStream` and `DigestOutputStream`. I think it's worth trying to use `startsWith("java.")`

Reply:
> You have to know that it is in the java.io package and it doesn't wrap another stream.

That is a good point. In the previous work on this override, we converged on the current implementation to not leak the internal byte[] to the target. It could be special cased for trusted targets but at the cost of auditing and complexity. So more thought needed on this, I don't think the current change can be integrated.

Reply:
> It could be special cased for trusted targets but at the cost of auditing and complexity. So more thought needed on this [...]

Could the target additionally be checked to see whether it is a `FilterOutputStream` and if so not be trusted?

Reply:
I've created a draft change for `OutputStream#write(ByteBuffer)`, and I have benchmarked several cases including `ByteArrayInputStream.transferTo(BufferedOutputStream(ByteArrayOutputStream))`, `ByteArrayInputStream.transferTo(DataOutputStream(FileOutputStream))`.

Benchmarks show there's improvement in both allocation rate and latency.

Read-only `ByteBuffers` can address poisoning concerns as non-jdk code can't peek into read-only arrays.

The `write(ByteBuffer)` approach supports cases like `CheckedOutputStream`, `DataOutputStream` which would be hard or impossible to optimize when passing naked byte arrays.

Here are the results: https://gist.github.com/vlsi/7f4411515a4f2dbb0925fffde92ccb1d
Here is the diff: https://github.com/vlsi/jdk/compare/ce108446ca1fe604ecc24bbefb0bf1c6318271c7...write_bytebuffer

@mkarg , @bplb , @AlanBateman , could you please review `OutputStream.write(ByteBuffer)` approach?

Reply:
@vlsi This is a very interesting solution 🤩, but it opens a number of questions! 🤔 As a start, here are mine:
* You propose a new public method (`OutputStream.write(ByteBuffer)`) as part of your solution. We need to discuss whether this is worth the effort to change all implementations (and to perceive acceptable performance, all custom authors need to optimize their custom classes, too). We also need to discuss whether we like the design choice that de facto the public API (not just the implementation) of the legacy IO domain (`OutputStream`) is now linked to the NIO domain (`ByteBuffer`) (which, IMHO, feels some kind of scary beyond `ChannelOutputStream`).
* Just thinking loudly: I wonder if we could simplify the solution, or if we could turn parts of it into some generic `byte[] readOnly(byte[])` utility.
* I would like to know how much faster the solution with `ByteBuffer` is compared to `Arrays.copyOfRange()`: Is it really so huge that it is worth the additional trouble?
* I would like to know how much slower the solution with `ByteBuffer` is compared to *skipping* `Arrays.copyOfRange()` for trusted cases (as you removed the skipping).
* I would like to know the performance of custom streams, because your default implementation is filling a temporary byte array with a Java-implemented byte-by-byte loop, which IMHO would be rather painful compared to hardware-supported `copyOrRange()`, and it does that even in case of *trusted* targets.
* @briangoetz Wouldn't we rather teach the Java language some day to provide native read-only arrays? That would be much faster, much better to read and implies less complex code for the end user.

Reply:
@mkarg , thank you for the review, I replied in JIRA to avoid mixing comments across different issues: https://bugs.openjdk.org/browse/JDK-8321271?focusedId=14634794&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14634794

Reply:
While frozen arrays are on our radar, this is a significant lift, so unlikely to be coming all that soon.

Reply:
I propose to separate this PR from Vladimir's proposal, so we can merge *this* PR first, and if we later see that Vladimir's changes are accepted, we can merge them *ontop*. Agreed everybody?